File: Ch. 12/Exhibit 12.1.2A/code/exhibit_12_1_2A.R
# =============================================================================
# Exhibit 12.1.2A: ATEs and Horowitz and Manski Bounds
# =============================================================================
# Compares default model with upper and lower bounds from Horowitz & Manski (2000).
# Column 1: Default model (no attrition adjustment)
# Column 2: Upper bound (best-case scenario for treatment effect)
# Column 3: Lower bound (worst-case scenario for treatment effect)
#
# Reference: Horowitz, J. L., & Manski, C. F. (2000). Nonparametric Analysis
# of Randomized Experiments with Missing Covariate and Outcome Data.
# Journal of the American Statistical Association, 95(449), 77-84.

# --- Setup -------------------------------------------------------------------
rm(list = ls())

# Load required libraries
library(tidyverse)
library(haven)
library(xtable)

# Automatically set working directory to the folder containing this script.
# Supports RStudio (Source/Knit), source(), and Rscript from the command line.
if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
} else if (length(sys.frames()) > 0 && !is.null(sys.frame(1)$ofile)) {
  setwd(dirname(sys.frame(1)$ofile))
} else if (!is.null(commandArgs(trailingOnly = FALSE))) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    setwd(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
}

# Define paths relative to script location
data_dir <- file.path(dirname(getwd()), "data")
output_dir <- file.path(dirname(getwd()), "output")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


# --- Parameters --------------------------------------------------------------
# Bounds for outcome variable (standardized cognitive test scores)
upper_bound <- 3
lower_bound <- -3


# --- Load and filter data ----------------------------------------------------
# Load synthetic data (for testing/demonstration)
data <- read_dta(file.path(data_dir, "unique_data_clean_main_synthetic.dta"))
# To use actual CHECC data, comment out the line above and uncomment the line below:
# data <- read_dta(file.path(data_dir, "unique_data_clean_main.dta"))

# Apply filters and create necessary variables
cleaned_data <- data %>%
  filter(year >= 2012 &
         (treatment == "control" | treatment == "prek") &
         kinderprep == 0 &
         late_randomized == 0 &
         has_cog_pre != 0) %>%
  mutate(
    # Create block variable (prioritize 2012 block, fallback to 2013)
    block = case_when(
      block_2012 != "" ~ block_2012,
      block_2013 != "" ~ block_2013
    ),
    # Set summer loss cognitive score to NA when not observed
    std_cog_sl = ifelse(has_cog_sl == 0, NA, std_cog_sl),
    # Treatment indicator: 1 if pre-K, 0 if control
    d_i = (treatment == "prek"),
    # Response indicator: 1 if outcome observed, 0 if attrited
    r_i = !is.na(std_cog_sl)
  )


# --- Filter for available cases ---------------------------------------------
# Default model uses only observed cases
filtered_data <- cleaned_data %>% filter(r_i == 1)


# --- Estimate models ---------------------------------------------------------
# Column 1: Default model (available cases only, no bounding)
model_default_hm <- lm(
  std_cog_sl ~ d_i,
  data = filtered_data
)

# Column 2: Horowitz & Manski upper bound
horowitz.manski.model.upper <- lm(
  std_cog_sl ~ d_i,
  data = cleaned_data %>% mutate(std_cog_sl = case_when(
    r_i == 0 & d_i == 1 ~ upper_bound,
    r_i == 0 & d_i == 0 ~ lower_bound,
    TRUE ~ std_cog_sl
  ))
)

# Column 3: Horowitz & Manski lower bound
horowitz.manski.model.lower <- lm(
  std_cog_sl ~ d_i,
  data = cleaned_data %>% mutate(std_cog_sl = case_when(
    r_i == 0 & d_i == 1 ~ lower_bound,
    r_i == 0 & d_i == 0 ~ upper_bound,
    TRUE ~ std_cog_sl
  ))
)


# --- Print results -----------------------------------------------------------
cat("\n", strrep("=", 80), "\n", sep = "")
cat("EXHIBIT 12.1.2A: ATEs and Horowitz and Manski Bounds\n")
cat(strrep("=", 80), "\n", sep = "")
cat(sprintf("%-20s %15s %15s %15s\n", "", "(1)", "(2)", "(3)"))
cat(sprintf("%-20s %15s %15s %15s\n", "", "Default", "H&M", "H&M"))
cat(sprintf("%-20s %15s %15s %15s\n", "", "Model", "Upper Bound", "Lower Bound"))
cat(strrep("-", 80), "\n", sep = "")

# Pre-K coefficient row
cat(sprintf("%-20s %15.3f %15.3f %15.3f\n",
            "Pre-K",
            coef(model_default_hm)["d_iTRUE"],
            coef(horowitz.manski.model.upper)["d_iTRUE"],
            coef(horowitz.manski.model.lower)["d_iTRUE"]))
cat(sprintf("%-20s (%13.3f) (%13.3f) (%13.3f)\n",
            "",
            summary(model_default_hm)$coefficients["d_iTRUE", "Std. Error"],
            summary(horowitz.manski.model.upper)$coefficients["d_iTRUE", "Std. Error"],
            summary(horowitz.manski.model.lower)$coefficients["d_iTRUE", "Std. Error"]))

# Constant row
cat(sprintf("%-20s %15.3f %15.3f %15.3f\n",
            "Constant",
            coef(model_default_hm)["(Intercept)"],
            coef(horowitz.manski.model.upper)["(Intercept)"],
            coef(horowitz.manski.model.lower)["(Intercept)"]))
cat(sprintf("%-20s (%13.3f) (%13.3f) (%13.3f)\n",
            "",
            summary(model_default_hm)$coefficients["(Intercept)", "Std. Error"],
            summary(horowitz.manski.model.upper)$coefficients["(Intercept)", "Std. Error"],
            summary(horowitz.manski.model.lower)$coefficients["(Intercept)", "Std. Error"]))

# Controls indicator row
cat(sprintf("%-20s %15s %15s %15s\n", "Controls", "No", "No", "No"))

# R-squared row
cat(sprintf("%-20s %15.3f %15.3f %15.3f\n",
            "R-squared",
            summary(model_default_hm)$r.squared,
            summary(horowitz.manski.model.upper)$r.squared,
            summary(horowitz.manski.model.lower)$r.squared))

# Observations row
cat(sprintf("%-20s %15d %15d %15d\n",
            "Observations",
            nobs(model_default_hm),
            nobs(horowitz.manski.model.upper),
            nobs(horowitz.manski.model.lower)))
cat(strrep("=", 80), "\n", sep = "")


# --- Save results to LaTeX ---------------------------------------------------
# Create summary table with regression results
results_table <- data.frame(
  Variable = c("Pre-K", "", "Constant", "", "Controls", "R-squared", "Observations"),
  `(1) Default Model` = c(
    sprintf("%.3f", coef(model_default_hm)["d_iTRUE"]),
    sprintf("(%.3f)", summary(model_default_hm)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(model_default_hm)["(Intercept)"]),
    sprintf("(%.3f)", summary(model_default_hm)$coefficients["(Intercept)", "Std. Error"]),
    "No",
    sprintf("%.3f", summary(model_default_hm)$r.squared),
    sprintf("%d", nobs(model_default_hm))
  ),
  `(2) H&M Upper Bound` = c(
    sprintf("%.3f", coef(horowitz.manski.model.upper)["d_iTRUE"]),
    sprintf("(%.3f)", summary(horowitz.manski.model.upper)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(horowitz.manski.model.upper)["(Intercept)"]),
    sprintf("(%.3f)", summary(horowitz.manski.model.upper)$coefficients["(Intercept)", "Std. Error"]),
    "No",
    sprintf("%.3f", summary(horowitz.manski.model.upper)$r.squared),
    sprintf("%d", nobs(horowitz.manski.model.upper))
  ),
  `(3) H&M Lower Bound` = c(
    sprintf("%.3f", coef(horowitz.manski.model.lower)["d_iTRUE"]),
    sprintf("(%.3f)", summary(horowitz.manski.model.lower)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(horowitz.manski.model.lower)["(Intercept)"]),
    sprintf("(%.3f)", summary(horowitz.manski.model.lower)$coefficients["(Intercept)", "Std. Error"]),
    "No",
    sprintf("%.3f", summary(horowitz.manski.model.lower)$r.squared),
    sprintf("%d", nobs(horowitz.manski.model.lower))
  ),
  check.names = FALSE
)

# Save to LaTeX
tex_file <- file.path(output_dir, "Exhibit_12_1_2A_r.tex")
print(xtable(results_table,
             caption = "ATEs and Horowitz and Manski Bounds",
             label = "tab:exhibit_12_1_2A"),
      file = tex_file,
      include.rownames = FALSE)
cat(sprintf("\n✓ Saved to: %s\n\n", tex_file))

# =============================================================================
# END OF EXHIBIT 12.1.2A
# =============================================================================

File: Ch. 12/Exhibit 12.1.2A/code/exhibit_12_1_2A.py
# =============================================================================
# Exhibit 12.1.2A: ATEs and Horowitz and Manski Bounds
# =============================================================================
# Compares default model with upper and lower bounds from Horowitz & Manski (2000).
# Column 1: Default model (no attrition adjustment)
# Column 2: Upper bound (best-case scenario for treatment effect)
# Column 3: Lower bound (worst-case scenario for treatment effect)
#
# Reference: Horowitz, J. L., & Manski, C. F. (2000). Nonparametric Analysis
# of Randomized Experiments with Missing Covariate and Outcome Data.
# Journal of the American Statistical Association, 95(449), 77-84.

from pathlib import Path
import pandas as pd
import statsmodels.api as sm


# --- Setup -------------------------------------------------------------------
# Automatically resolve paths relative to this script's location.
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR.parent / "data"
OUTPUT_DIR = SCRIPT_DIR.parent / "output"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


# --- Parameters --------------------------------------------------------------
# Bounds for outcome variable (standardized cognitive test scores)
UPPER_BOUND = 3
LOWER_BOUND = -3


# --- Load and filter data ----------------------------------------------------
# Load synthetic data (for testing/demonstration)
data_12_1_2A = pd.read_stata(DATA_DIR / "unique_data_clean_main_synthetic.dta")
# To use actual CHECC data, comment out the line above and uncomment the line below:
# data_12_1_2A = pd.read_stata(DATA_DIR / "unique_data_clean_main.dta")


data_12_1_2A['year'] = pd.to_numeric(data_12_1_2A['year'], errors='coerce')

# Apply filters: year >= 2012, treatment in {control, prek}, exclude kinderprep
# and late randomized, require baseline cognitive score
cleaned_data_12_1_2A = data_12_1_2A[
    (data_12_1_2A['year'] >= 2012) &
    ((data_12_1_2A['treatment'] == 'control') |
     (data_12_1_2A['treatment'] == 'prek')) &
    (data_12_1_2A['kinderprep'] == 0) &
    (data_12_1_2A['late_randomized'] == 0) &
    (data_12_1_2A['has_cog_pre'] != 0)
].copy()

# Create block variable (prioritize 2012 block, fallback to 2013)
cleaned_data_12_1_2A['block'] = cleaned_data_12_1_2A.apply(
    lambda row: row['block_2012'] if row['block_2012'] != "" else row['block_2013'],
    axis=1
)

# Set summer loss cognitive score to NA when not observed
cleaned_data_12_1_2A['std_cog_sl'] = cleaned_data_12_1_2A['std_cog_sl'].where(
    cleaned_data_12_1_2A['has_cog_sl'] != 0, pd.NA
)

# Treatment indicator: 1 if pre-K, 0 if control
cleaned_data_12_1_2A['d_i'] = (cleaned_data_12_1_2A['treatment'] == "prek").astype(int)

# Response indicator: 1 if outcome observed, 0 if attrited
cleaned_data_12_1_2A['r_i'] = cleaned_data_12_1_2A['std_cog_sl'].notna().astype(int)


# --- Create bounding datasets ------------------------------------------------
# Upper bound: Assign best outcome to treatment attritors, worst to control attritors
# This maximizes the treatment effect estimate
upper_bound_data = cleaned_data_12_1_2A.copy()
upper_bound_data.loc[
    (upper_bound_data['r_i'] == 0) & (upper_bound_data['d_i'] == 1),
    'std_cog_sl'
] = UPPER_BOUND
upper_bound_data.loc[
    (upper_bound_data['r_i'] == 0) & (upper_bound_data['d_i'] == 0),
    'std_cog_sl'
] = LOWER_BOUND

# Lower bound: Assign worst outcome to treatment attritors, best to control attritors
# This minimizes the treatment effect estimate
lower_bound_data = cleaned_data_12_1_2A.copy()
lower_bound_data.loc[
    (lower_bound_data['r_i'] == 0) & (lower_bound_data['d_i'] == 1),
    'std_cog_sl'
] = LOWER_BOUND
lower_bound_data.loc[
    (lower_bound_data['r_i'] == 0) & (lower_bound_data['d_i'] == 0),
    'std_cog_sl'
] = UPPER_BOUND


# --- Filter for available cases ---------------------------------------------
# Default model uses only observed cases
filtered_data_12_1_2A = cleaned_data_12_1_2A[cleaned_data_12_1_2A['r_i'] == 1]


# --- Estimate models ---------------------------------------------------------
# Column 1: Default model (available cases only, no bounding)
model_default_hm = sm.OLS.from_formula(
    'std_cog_sl ~ d_i',
    data=filtered_data_12_1_2A
).fit()

# Column 2: Horowitz & Manski upper bound
model_hm_upper = sm.OLS.from_formula(
    'std_cog_sl ~ d_i',
    data=upper_bound_data
).fit()

# Column 3: Horowitz & Manski lower bound
model_hm_lower = sm.OLS.from_formula(
    'std_cog_sl ~ d_i',
    data=lower_bound_data
).fit()


# --- Print results -----------------------------------------------------------
print("\n" + "=" * 80)
print("EXHIBIT 12.1.2A: ATEs and Horowitz and Manski Bounds")
print("=" * 80)
print(f"{'':20} {'(1)':>15} {'(2)':>15} {'(3)':>15}")
print(f"{'':20} {'Default':>15} {'H&M':>15} {'H&M':>15}")
print(f"{'':20} {'Model':>15} {'Upper Bound':>15} {'Lower Bound':>15}")
print("-" * 80)

# Pre-K coefficient row
print(f"{'Pre-K':<20} {model_default_hm.params['d_i']:>15.3f} "
      f"{model_hm_upper.params['d_i']:>15.3f} "
      f"{model_hm_lower.params['d_i']:>15.3f}")
print(f"{'':20} ({model_default_hm.bse['d_i']:>13.3f}) "
      f"({model_hm_upper.bse['d_i']:>13.3f}) "
      f"({model_hm_lower.bse['d_i']:>13.3f})")

# Constant row
print(f"{'Constant':<20} {model_default_hm.params['Intercept']:>15.3f} "
      f"{model_hm_upper.params['Intercept']:>15.3f} "
      f"{model_hm_lower.params['Intercept']:>15.3f}")
print(f"{'':20} ({model_default_hm.bse['Intercept']:>13.3f}) "
      f"({model_hm_upper.bse['Intercept']:>13.3f}) "
      f"({model_hm_lower.bse['Intercept']:>13.3f})")

# Controls indicator row
print(f"{'Controls':<20} {'No':>15} {'No':>15} {'No':>15}")

# R-squared row
print(f"{'R-squared':<20} {model_default_hm.rsquared:>15.3f} "
      f"{model_hm_upper.rsquared:>15.3f} "
      f"{model_hm_lower.rsquared:>15.3f}")

# Observations row
print(f"{'Observations':<20} {int(model_default_hm.nobs):>15} "
      f"{int(model_hm_upper.nobs):>15} "
      f"{int(model_hm_lower.nobs):>15}")
print("=" * 80)


# --- Save results to LaTeX ---------------------------------------------------
# Create summary table with regression results
results_table = pd.DataFrame({
    'Variable': ['Pre-K', '', 'Constant', '', 'Controls', 'R-squared', 'Observations'],
    '(1) Default Model': [
        f"{model_default_hm.params['d_i']:.3f}",
        f"({model_default_hm.bse['d_i']:.3f})",
        f"{model_default_hm.params['Intercept']:.3f}",
        f"({model_default_hm.bse['Intercept']:.3f})",
        'No',
        f"{model_default_hm.rsquared:.3f}",
        f"{int(model_default_hm.nobs)}"
    ],
    '(2) H&M Upper Bound': [
        f"{model_hm_upper.params['d_i']:.3f}",
        f"({model_hm_upper.bse['d_i']:.3f})",
        f"{model_hm_upper.params['Intercept']:.3f}",
        f"({model_hm_upper.bse['Intercept']:.3f})",
        'No',
        f"{model_hm_upper.rsquared:.3f}",
        f"{int(model_hm_upper.nobs)}"
    ],
    '(3) H&M Lower Bound': [
        f"{model_hm_lower.params['d_i']:.3f}",
        f"({model_hm_lower.bse['d_i']:.3f})",
        f"{model_hm_lower.params['Intercept']:.3f}",
        f"({model_hm_lower.bse['Intercept']:.3f})",
        'No',
        f"{model_hm_lower.rsquared:.3f}",
        f"{int(model_hm_lower.nobs)}"
    ]
})

# Save to LaTeX
tex_file = OUTPUT_DIR / "Exhibit_12_1_2A_python.tex"
results_table.to_latex(tex_file, index=False, escape=False,
                       caption="ATEs and Horowitz and Manski Bounds",
                       label="tab:exhibit_12_1_2A")
print(f"\n✓ Saved to: {tex_file}\n")

# =============================================================================
# END OF EXHIBIT 12.1.2A
# =============================================================================
